src/HtmlReader/architecture.md

```diff
@@ -0,0 +1,78 @@
+
+- `InactiveState`: The hook was not provided any arguments, and thus it is inactive.
+- `FetchingResourceState`: We are fetching a resource, and thus are "loading".
+- `ResourceFetchErrorState`: We have encountered an error while fetching the resource.
+- `RenderingIframeState`: We have the resource, and are now waiting for the iframe to finish rendering so that we have a reference to it. We get this reference via the `ref` property on the `<iframe>` tag. This is a very short state as it only exists during the single render cycle between when the resource finishes fetching and the iframe renders for the first time. Note that rendering the iframe element does not mean the iframe has rendered its children, that is separate.
+- `LoadingIframeState`: We have rendered the iframe once, and are now waiting for it to fire off it's own `window.onLoad` function indicating that it has loaded all the html in the resource string we passed it.
+- `NavigatingState`: The iframe has finished loading, and now we are waiting for our `useEffect` hook to complete any inter-resource navigation that might be necessary depending on the current `state.location`. For example, scrolling the user to a specific page.
+- `ReadyState`: We have now navigated the user to a specific page, and are now finished with our work.
+
+The definitions for these states can be found in `./types.ts`. Note that state transitions should happen in the order listed above, with the exception that it is possible to go backwards occasionally. For example, from the `ReadyState`, we may go back to `FetchingResourceState` if the user switches to a new resource, or we might simply go back to the `NavigatingState` if they navigate to a new location in the same resource.
+
+We provide the text value `state.state` as a string describing the state. This makes it very easy to perform type narrowing. If `state.state === 'READY'`, typescript knows that `state.iframe` and `state.resource`, and more all exist.
+
+### Reducer
+
+It is recommended to look through the `./reducer` file to see how various actions (found defined in `./types.ts`) update the sate.
+
+## Fetching resources
+
+When the state is `FetchingResourceState`, we need to actually go out and fetch the resource referenced in `state.location.href`. We do this in the `useResource` hook. When we transition to a `FetchingResourceState`, that hook will call `getContent` to get the resource. It then will dispatch a `RESOURCE_FETCH_SUCCESS` action, which will transition the state to `RenderingIframeState`.
+
+### Updating the resource
+
+After fetching the resource, there are a couple things we need to do to it before we can render it:
+
+- Set a `<base>` element in the `<head>` with the url of the resource. This tells the iframe where relative urls within the resource are relative to.
+- Inject any injectables. For example, we need to add the Readium-CSS stylesheets to the `<head>`.
```

Comment:
Is CSS the only injectible?  What other injectibles should we consider? Do these have implications for the rest of the update-resource process?

Reply:
No, another injectable is JS or, theoretically, any element that can go in the head of the document. Typically, we should be prepared for CSS and JS injections. They don't have implications because we don't support _changing_ the injectables after initialization. We don't need that, and I think if someone wanted it, we would let them do it with some kind of API. But for now, no, the injectables are injected only when the resource is fetched.